core/unit_test/tools/TestLogicalSpaces.hpp

```diff
@@ -0,0 +1,201 @@
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+#include <iostream>
+#include <gtest/gtest.h>
+#include "Kokkos_Core.hpp"
+
+#include <impl/Kokkos_Stacktrace.hpp>
+
+namespace Test {
+
+void debug_print(const Kokkos_Profiling_SpaceHandle hand, const char* name,
+                 const void* ptr, const size_t size) {
+  std::cout << "Alloc: " << hand.name << ", [" << name << "," << ptr << "] "
+            << size << std::endl;
+}
+void debug_dealloc(const Kokkos_Profiling_SpaceHandle hand, const char* name,
+                   const void* ptr, const size_t size) {
+  std::cout << "Dealloc: " << hand.name << ", [" << name << "," << ptr << "] "
+            << size << std::endl;
+}
+
+void fail_on_event(const Kokkos::Profiling::SpaceHandle, const char*,
+                   const void*, const uint64_t) {
+  ASSERT_TRUE(false) << "Unexpected memory event";
+}
+
+void expect_no_events() {
+  Kokkos::Tools::Experimental::set_allocate_data_callback(&fail_on_event);
+  Kokkos::Tools::Experimental::set_deallocate_data_callback(&fail_on_event);
+}
+
+std::string expected_view_name;
+std::string expected_space_name;
+std::string error_message;
```

Comment:
Why do these need to be global variables?

Reply:
Because they're used in Profiling callbacks, so they can't be captured. There are slight variations on this trick, but they're all fundamentally the same as what I'm doing here

Reply:
> they can't be captured

I am not following.  Please elaborate.

Reply:
So, the natural way to write this would be something like

```c++
void expect_allocation_event(const std::string evn, const std::string esn,
                             const std::string em) {
    Kokkos::Tools::Experimental::set_allocate_data_callback(
      [](const Kokkos_Profiling_SpaceHandle hand, const char* name, const void*,
         const uint64_t) {
        ASSERT_EQ(std::string(hand.name), esn)
            << error_message << " (bad handle)";
        ASSERT_EQ(std::string(name), evn)
            << error_message << " (bad view name)";
        expect_no_events();
      });
}
```

Problem being, that would lead to `evn` and the like being captured in that lambda. But if the lambda captures, it can't be cast to a function pointer, and `set_allocate_data_callback` takes a *function pointer* (excuse the emphasis, it's just really easy to misread that as "functor" in the middle of Kokkos) as an argument. So the lambda being defined here can't capture things. But a lambda doesn't capture things that are declared at top-level scope (more formally it's something to do with `odr` usage), so having the lambda use globals, and set those globals just above, is kosher.

Reply:
Yeah, w/o an extra parameter, store a std::function, etc., a global/static is required somewhere.

(FYI:  lambdas don't capture globals/statics because they have a fixed address in memory that is just accessed directly.)